migrations/versions/3d3a1981593c_populate_fields_from_systems_to_.py

```diff
@@ -0,0 +1,88 @@
+"""populate fields from systems to performance profile
+
+Revision ID: 3d3a1981593c
+Revises: 015d7d1fea52
+Create Date: 2022-03-23 23:45:14.430980
+
+"""
+from migrations.session_helpers import session
+from sqlalchemy import func, DateTime, Integer, String, Column
+from sqlalchemy.dialects.postgresql import UUID, JSONB
+from sqlalchemy.ext.declarative import declarative_base
+from ros.lib.utils import get_or_create
+
+# revision identifiers, used by Alembic.
+revision = '3d3a1981593c'
+down_revision = '015d7d1fea52'
+branch_labels = None
+depends_on = None
+
+Base = declarative_base()
+
+
+class PerformanceProfile(Base):
+    __tablename__ = "performance_profile"
+    system_id = Column(Integer, primary_key=True)
+    report_date = Column(DateTime, primary_key=True)
+    rule_hit_details = Column(JSONB)
+    state = Column(String(25))
+    number_of_recommendations = Column(Integer)
+    operating_system = Column(JSONB)
+
+
+class System(Base):
+    __tablename__ = 'systems'
+    id = Column(Integer, primary_key=True)
+    inventory_id = Column(UUID(as_uuid=True), nullable=False)
+    rule_hit_details = Column(JSONB)
+    state = Column(String(25))
+    number_of_recommendations = Column(Integer)
+    operating_system = Column(JSONB)
+
+
+def populate_fields_data_to_performance_profile():
+    with session() as session_obj:
+        last_reported = (
+            session_obj.query(
+                PerformanceProfile.system_id,
+                func.max(PerformanceProfile.report_date).label('max_date')
+            ).group_by(PerformanceProfile.system_id).subquery()
+        )
+
+        query = (
+            session_obj.query(PerformanceProfile, System)
+            .join(
+                last_reported, (
+                    last_reported.c.max_date == PerformanceProfile.report_date
+                ) & (
+                    PerformanceProfile.system_id == last_reported.c.system_id))
+            .join(System, System.id == last_reported.c.system_id)
+        )
+
+        pp_results = query.all()
+        for record in pp_results:
+            try:
+                get_or_create(
+                    session_obj, PerformanceProfile,
+                    ['system_id', 'report_date'],
+                    system_id=record.System.id,
+                    report_date=record.PerformanceProfile.report_date,
+                    operating_system=record.System.operating_system,
+                    state=record.System.state,
+                    rule_hit_details=record.System.rule_hit_details,
+                    number_of_recommendations=record.System.number_of_recommendations
+                )
+            except Exception as err:
+                print(
+                    'Failed to update performance_profile with '
+                    f'system_id={record.System.id} and error={err}'
+                )
+                raise
+
+
+def upgrade():
+    populate_fields_data_to_performance_profile()
+
+
+def downgrade():
+    pass
```

Comment:
@r14chandra - it is good idea to implement `downgrade` method but here I didn't find any scenario where it will be useful. Do you think of any scenario where downgrade for this migration will help?

Reply:
There might be a case on `local`, where we upgrade db to a specific revision and then downgrade. Db might not be consistent.

Reply:
That's what I would like to understand in which scenario one wants to upgrade & then downgrade this specific version.

The two scenarios, I would think of are that 
- One wants to change logic from upgrade and rerun. This migration is to populate values into fields even if someone changes the logic -> downgrade -> rerun, it will simply run as it is.
- One wants to remove existing population then one can downgrade db till version `015d7d1fea52` and just rerun upgrade db to `015d7d1fea52`. Will it cover scenario that you mentioned?

I don't have strong opinion on this but I think it is OK not to have downgrade method for this migration.

Reply:
Thank you @r14chandra for highlighting your concern for local setup where one wants to switch between current ongoing changes plus historical data migrations testing on local.
Using DB backup + restore, you will be able to continue your ongoing change smoothly without disturbing DB schema.

Reply:
@r14chandra IIUC python manage.py db `downgrade` should already be present on alembic level. I have used it for cleanups.

```
(ros-backend) ➜  ros-backend git:(RHIROS-573-fix-db-schema-pp-history) ✗ python manage.py db downgrade
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade 77fdc857c94f -> ccf59577aca2, alter pk on performance profile
```

@kgaikwad 
Scenario: 

I have decided `x` field to have `int` datatype in respective model, the I realize it should be `str`.
Instead of pushing both resulting migrations to a PR I can downgrade, manually delete the `int` datatype migration file and include just the necessary one.